ggml/src/ggml.c

```diff
@@ -6657,43 +6625,44 @@ struct gguf_context * gguf_init_from_file(const char * fname, struct gguf_init_p
                             case GGUF_TYPE_FLOAT64:
                             case GGUF_TYPE_BOOL:
                                 {
-                                    // prevent from integer overflow in the malloc below
+                                    // prevent integer overflow in the malloc below
                                     if (kv->value.arr.n >= SIZE_MAX/gguf_type_size(kv->value.arr.type)) {
                                         fprintf(stderr, "%s: array size is too large (%" PRIu64 ")\n", __func__, kv->value.arr.n);
                                         fclose(file);
                                         gguf_free(ctx);
                                         return NULL;
                                     }
 
-                                    kv->value.arr.data = calloc(kv->value.arr.n, gguf_type_size(kv->value.arr.type));
+                                    const size_t nbytes = kv->value.arr.n * gguf_type_size(kv->value.arr.type);
+                                    kv->value.arr.data = malloc(nbytes);
```

Comment:
Do not change `calloc` to `malloc`, it is done that way to reduce the chance of security issues.

Reply:
Should we also add a function like `ggml_crealloc` (since to my knowledge no equivalent C function exists)?

Reply:
I think what we should do is rewrite this code in C++. This code is too sensitive to security issues, and C does not provide the necessary abstractions to be able to reason about this code such that we can be reasonably confident that it is memory safe.

Reply:
To answer your question, we should return a failure rather than crashing when a memory allocation (or anything else) fails in the GGUF loader. If that's not possible, I guess crashing is better than nothing, but not really by much.

Reply:
I've already made myself familiar with how the code works anyways so I would be fine with rewriting it as C++. Are there things in particular that would be important to consider?

Reply:
They key aspect would be using RAII to handle all the resource allocations. The objects should always be in a consistent state such as the destructor call be called at any time to destroy them if necessary. Use exceptions if fatal errors in inner functions cannot be propagated safely in other way. Generally should use C++ containers and never rely on manual memory management. All memory accesses should be bounds-checked as well, e.g. there are some known issues because the KV array reading functions do not check bounds.